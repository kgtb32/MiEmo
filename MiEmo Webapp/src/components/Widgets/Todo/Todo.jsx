import React, { useState, useRef, useEffect } from 'react'
import { DataTable } from 'primereact/datatable'
import { Column } from 'primereact/column'
import { InputText } from 'primereact/inputtext'
import { Toast } from 'primereact/toast'
import { Checkbox } from 'primereact/checkbox'
import styled from 'styled-components'
import { Button } from 'primereact/button/'
import Container from 'react-bootstrap/Container'
import Row from 'react-bootstrap/Row'

const Todo = () => {
	const [task, setTask] = useState(JSON.parse(localStorage.getItem('taskwidgetdata')) ?? [])
	const [editingRows, setEditingRows] = useState({})
	const toast = useRef(null)
	const [checked, setChecked] = useState(false)
	const [newtaskname, setNewTaskName] = useState()
	const [error, setError] = useState(false)

	useEffect(() => {
		localStorage.setItem('taskwidgetdata', JSON.stringify(task))
	}, [task])

	const onRowEditComplete = e => {
		setTask(task.map(item => (item.id !== e.newData.id ? item : e.newData)))
	}

	const onRowEditDelete = e => {
		setTask(task.filter(item => item.id !== task[e.index].id))
	}

	const onRowEditChange = e => {
		setEditingRows(e.data)
	}

	const textEditor = options => {
		return <InputText type="text" value={options.value} onChange={e => options.editorCallback(e.target.value)} />
	}

	const checkEditor = options => {
		return (
			<Checkbox
				inputId="binary"
				checked={checked}
				onChange={e => {
					setChecked(e.checked)
					options.editorCallback(e.checked)
				}}
			/>
		)
	}

	const checkBodyTemplate = rowData => {
		switch (rowData.check) {
			case true:
				return 'Terminé'

			case false:
				return 'Pas Terminé'
		}
	}

	const addTask = () => {
		const taskwidgetdata = JSON.parse(localStorage.getItem('taskwidgetdata'))
		let _id = taskwidgetdata && taskwidgetdata.length > 0 ? Math.max(...task.map(value => value.id)) + 1 : 1
		if (newtaskname) {
			setTask([
				...task,
				{
					id: _id,
					taskname: newtaskname,
					check: false,
				},
			])
			setError(false)
		} else {
			setError(true)
			toast.current.show({ severity: 'warn', summary: 'Error', detail: 'Remplisser les champs', life: 3000 })
		}
	}

	return (
		<div className="datatable-editing-demo overflow-auto h-100">
			<Toast ref={toast} />

			<div className="card">
				<h5>Tâche</h5>
				<InputTaskContainer>
					<Button
						icon="pi pi-plus"
						className="p-button-rounded p-button-secondary"
						aria-label="Bookmark"
						onClick={() => addTask()}
					/>
					<InputTask>
						<h5>Nom de la Tâche</h5>
						<Container>
							<Row>
								<InputText
									value={newtaskname}
									onChange={e => setNewTaskName(e.target.value)}
									className={error ? 'p-invalid block' : ''}
								/>
							</Row>
						</Container>
						{error ? <small className="p-error block">Entrez une tâche</small> : ''}
					</InputTask>
				</InputTaskContainer>
				<div className="p-fluid">
					<DataTable
						value={task}
						editMode="row"
						dataKey="id"
						editingRows={editingRows}
						onRowEditChange={onRowEditChange}
						onRowEditComplete={onRowEditComplete}
						onRowEditCancel={onRowEditDelete}
						responsiveLayout="scroll"
					>
						<Column
							field="taskname"
							header="Tâche"
							editor={options => textEditor(options)}
							style={{ width: '20%' }}
						></Column>
						<Column
							field="check"
							header="Check"
							editor={options => checkEditor(options)}
							body={checkBodyTemplate}
							style={{ width: '20%' }}
						></Column>
						<Column
							rowEditor
							headerStyle={{ width: '10%', minWidth: '8rem' }}
							bodyStyle={{ textAlign: 'center' }}
						></Column>
					</DataTable>
				</div>
			</div>
		</div>
	)
}

const InputTaskContainer = styled.div`
	display: flex;
	flex-direction: row;
	flex-wrap: wrap;
	align-items: center;
	justify-content: space-evenly;
	padding: 20px;
`

const InputTask = styled.div`
	display: flex;
	flex-direction: column;
	align-items: center;
`
export default Todo
