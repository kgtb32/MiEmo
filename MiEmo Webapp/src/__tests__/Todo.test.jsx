import React from 'react'
import { render, unmountComponentAtNode } from 'react-dom'
import Todo from '../components/Widgets/Todo/Todo'
import { act } from 'react-dom/test-utils'
import { shallow, configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

let container = null
beforeEach(() => {
	// setup a DOM element as a render target
	container = document.createElement('div')
	document.body.appendChild(container)
	act(() => {
		render(<Todo />, container)
	})
	configure({ adapter: new Adapter() })
})

afterEach(() => {
	// cleanup on exiting
	unmountComponentAtNode(container)
	container.remove()
	container = null
})

describe('rendering test', () => {
	// verify that the Todo HTML has not changed between git merges
	it('It should match the snapshot HTML for Todo', () => {
		const container = shallow(<Todo />)
		expect(container.html()).toMatchSnapshot()
	})

	it('It should contain all static text when firts render', () => {
		expect(container.textContent).toContain('Nom de la T창che')
		expect(container.textContent).toContain('T창che')
		expect(container.textContent).toContain('Check')
		expect(container.textContent).toContain('No available options')
	})

	it('Input should be empty', () => {
		const inputValue = document.getElementsByTagName('input')[0].value
		expect(inputValue).toEqual('')
		const button = document.querySelector('button')
		expect(button.disabled).toEqual(false)
	})

	it('Button should be enable', () => {
		const button = document.querySelector('button')
		expect(button.disabled).toEqual(false)
	})
})

describe('fonctional test', () => {
	it('Local storage should not contains task when first render', () => {
		expect(JSON.parse(localStorage.getItem('taskwidgetdata'))).toEqual([])
	})

	it('Task should not be added update when input is not completed', () => {
		const button = document.getElementsByClassName('p-button-rounded p-button-secondary')[0]
		button.dispatchEvent(new MouseEvent('click', { bubbles: true }))
		expect(container.textContent).toContain('Error')
		expect(container.textContent).toContain('Remplisser les champs')
		expect(container.textContent).toContain('Nom de la T창che')
		expect(container.textContent).toContain('T창che')
		expect(container.textContent).toContain('Check')
		expect(container.textContent).toContain('No available options')
	})

	it('It should contain Task when adding task', () => {
		//set value on input
		const input = document.getElementsByClassName('p-inputtext p-component')[0]
		const nativeInputValueSetter = Object.getOwnPropertyDescriptor(window.HTMLInputElement.prototype, 'value').set
		nativeInputValueSetter.call(input, 'My task')
		const ev2 = new Event('input', { bubbles: true })
		input.dispatchEvent(ev2)

		const button = document.getElementsByClassName('p-button-rounded p-button-secondary')[0]
		button.dispatchEvent(new MouseEvent('click', { bubbles: true }))
		expect(container.textContent).toContain('My task')
		localStorage.clear()
	})

	it('It should not contain No available options when adding task', () => {
		//set value on input
		const input = document.getElementsByClassName('p-inputtext p-component')[0]
		const nativeInputValueSetter = Object.getOwnPropertyDescriptor(window.HTMLInputElement.prototype, 'value').set
		nativeInputValueSetter.call(input, 'My task')
		const ev2 = new Event('input', { bubbles: true })
		input.dispatchEvent(ev2)

		const button = document.getElementsByClassName('p-button-rounded p-button-secondary')[0]
		button.dispatchEvent(new MouseEvent('click', { bubbles: true }))
		expect(container.textContent).not.toContain('No available options')
		localStorage.clear()
	})

	it('Local storage should contain Task when adding task', () => {
		//set value on input
		const input = document.getElementsByClassName('p-inputtext p-component')[0]
		const nativeInputValueSetter = Object.getOwnPropertyDescriptor(window.HTMLInputElement.prototype, 'value').set
		nativeInputValueSetter.call(input, 'My task')
		const ev2 = new Event('input', { bubbles: true })
		input.dispatchEvent(ev2)

		const button = document.getElementsByClassName('p-button-rounded p-button-secondary')[0]
		button.dispatchEvent(new MouseEvent('click', { bubbles: true }))

		expect(JSON.parse(localStorage.getItem('taskwidgetdata'))[0].taskname).toBe('My task')
		localStorage.clear()
	})
})
