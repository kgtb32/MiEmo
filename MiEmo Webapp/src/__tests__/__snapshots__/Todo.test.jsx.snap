// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`view testing It should match the snapshot HTML for Todo 1`] = `
<DocumentFragment>
  <div
    class="sc-dkrFOg bmZLDp datatable-editing-demo overflow-auto h-100"
    data-testid="test-todo"
  >
    <div
      class="p-toast p-component p-toast-top-right"
    >
      <div />
    </div>
    <div
      class="card font-bold"
    >
      <div
        class="d-flex"
      >
        <div
          class="w-100"
        >
          <div
            class="w-100"
            placeholder="Entrez une tâche"
          >
            <input
              class="p-inputtext p-component w-100"
              placeholder="Entrez une tâche"
              value=""
            />
          </div>
        </div>
        <button
          aria-label="Bookmark"
          class="p-button p-component sc-gswNZR jPIqlE p-button-icon-only p-disabled"
          data-testid="test-buttonadd"
          disabled=""
        >
          <span
            class="p-button-icon p-c pi pi-plus"
          />
          <span
            class="p-button-label p-c"
          >
             
          </span>
        </button>
      </div>
      <div
        class="p-fluid"
      >
        <div
          class="p-datatable p-component p-datatable-responsive-scroll"
          data-scrollselectors=".p-datatable-wrapper"
          pr_id_1=""
        >
          <div
            class="p-datatable-wrapper"
          >
            <table
              class="p-datatable-table"
              role="table"
            >
              <thead
                class="p-datatable-thead"
              >
                <tr
                  role="row"
                >
                  <th
                    class=""
                    role="columnheader"
                    style="width: 20%;"
                  >
                    <div
                      class="p-column-header-content"
                    >
                      <span
                        class="p-column-title"
                      >
                        Tâche
                      </span>
                    </div>
                  </th>
                  <th
                    class=""
                    role="columnheader"
                    style="width: 20%;"
                  >
                    <div
                      class="p-column-header-content"
                    >
                      <span
                        class="p-column-title"
                      >
                        Check
                      </span>
                    </div>
                  </th>
                  <th
                    class=""
                    role="columnheader"
                    style="width: 10%; min-width: 8rem;"
                  >
                    <div
                      class="p-column-header-content"
                    >
                      <span
                        class="p-column-title"
                      />
                    </div>
                  </th>
                </tr>
              </thead>
              <tbody
                class="p-datatable-tbody"
              >
                <tr
                  class="p-datatable-emptymessage"
                  role="row"
                >
                  <td
                    colspan="3"
                    role="cell"
                  >
                    No results found
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
